### DISCUSSION TITLE
optimistic locking

### DISCUSSION THREAD

**Discussion by jkdoshi** [pocketbase/pocketbase#5178] [Q&A] [answered]
How to handle optimistic locking in PocketBase?

Let me explain.
Assume there's a PB server and a JS SPA app client (using js-sdk):
- The user opens the same record in an edit form in two browser tabs. They both started with the same copy of the original record.
- User saves (form-submits) the first tab and then the second one.
- The second submit overwrites the first one.

How to prevent that?

An optimistic lock implementation would simply check that the "updated" timestamp it is overwriting is same as the updated timestamp it had when the record was loaded. If the updated timestamp is different, then it means the record has been changed/updated since the current page loaded it. Something like ...

```sql
UPDATE foo SET ...
WHERE id = {:id}
AND updated = {:updated}
```

The "AND update = {:updated}" is what enforces optimistic locking, causing the UPDATE to fail if someone has already updated that record since you last loaded it.

How can I enforce the above?

Currently, in my app, two users or two tabs are overwriting each other (last write wins).

**Comment by ganigeorgiev** [comment, marked as answer]
I'm not entirely sure if that is what you are looking for, but technically nothing prevents you to submit an additional query or body date value with the request and compare it with the `updated` field in the API rules:

```js
// allow updates only if the submitted someParam date is after the record updated
updated < @request.query.someParam
```
(_the someParam date can be created for example on form initialization_)

Alternatively, if you just want to target a single user within the browser you can make use of the [`BroadcastChannel`](https://developer.mozilla.org/en-US/docs/Web/API/Broadcast_Channel_API) browser api to sync the forms content and state between tabs (_at the moment we use it only for syncing the Svelte collections store in the Admin UI_).

**Reply by jkdoshi** [reply]
As always, amazingly quick reply @ganigeorgiev .

I have tested this approach, and it works. Thanks!

HOWEVER ... it requires me to make two changes for ALL my collections, and ALL the PocketBase.update js-sdk update calls.

- update the "update" API rules of all collections to include `updated = @request.query.updated`
- update all PocketBase.update calls to ... `pb.update(id, data, { query: {updated} })`

That seems like a big global change for the application. Plus, it seems that most web-application should include the optimistic locking feature anyways. So, what if you include it for free (turned ON or OFF by some connection option) in the PocketBase server itself?

- the js-sdk client sends the `updated` field with all PUT/PATCH requests
- the server includes `updated = {:updated}` clause in all `UPDATE` queries

I know it's a big change for you too. But I think in end it will make PocketBase more robust.

My own immediate need is taken care of. So thank you for that.

**Reply by ganigeorgiev** [reply]
No, sorry. There are no plans for this (_with the refactoring the `created`/`updated` fields are also optional_).

**Comment by tobiasbenkner** [comment]
Hi @jkdoshi,

I just encountered the same problem and I solved it using the hooks in golang. In my case you don't need to touch the api filters.

**Solution**
- If you want to have optimistic locking activated, you can create a version attribute with the data type number in your collection
- When you add a dataset, the value 1 is automatically entered in the version field
- When updating the data set, the current version must be included in the payload
  - if the version attribute is missing or an incorrect version is transmitted, a 412 or 428 error code is returned. You can react better to this in the frontend. (In your solution i guess you will get a 404 error. Maybe there was something wrong with your payload)
  - if successful, the version is automatically incremented
- This approach also works if the create and update attributes do not exist
- I have not yet tested what happens when 2 people update at the same time. There could still be a race condition

```golang
app.OnRecordBeforeCreateRequest().Add(func(e *core.RecordCreateEvent) error {

	record := e.Record
	if field := record.Collection().Schema.GetFieldByName("version"); field == nil {
		// no version - ignore checking
		return nil
	}

	record.Set("version", 1)
	return nil
})

app.OnRecordBeforeUpdateRequest().Add(func(e *core.RecordUpdateEvent) error {
	record := e.Record

	if field := record.Collection().Schema.GetFieldByName("version"); field == nil {
		// no version - ignore checking
		return nil
	}

	version := record.GetInt("version")
	if version == 0 {
		return echo.NewHTTPError(http.StatusPreconditionRequired, "the version attribute is required for this operation and is missing from the request.")
	}

	collectionName := record.Collection().Name
	id := record.Id
	foundRecord, err := app.Dao().FindRecordById(collectionName, id)
	if err != nil {
		return echo.NewHTTPError(http.StatusNotFound, "cannot find record with id "+id+" in collection "+collectionName)
	}

	foundVersion := foundRecord.GetInt("version")

	if version != foundVersion {
		return echo.NewHTTPError(http.StatusPreconditionFailed, "the version attribute does not match the current version of the resource. Please reload and try again.")
	}

	record.Set("version", version+1)
	return nil
})
```

**Reply by tobiasbenkner** [reply]
my previous code cannot handle parallel requests. to fix it you have to use a transaction. a transaction will be executed atomically. this way you can avoid parallel access. the first request wins and the other requests are blocked, as it should be.

```golang
app.OnRecordBeforeCreateRequest().Add(func(e *core.RecordCreateEvent) error {

	record := e.Record
	if field := record.Collection().Schema.GetFieldByName("version"); field == nil {
		// no version - ignore checking
		return nil
	}

	record.Set("version", 1)
	return nil
})

app.OnRecordBeforeUpdateRequest().Add(func(e *core.RecordUpdateEvent) error {
	record := e.Record

	if field := record.Collection().Schema.GetFieldByName("version"); field == nil {
		// no version - ignore checking
		return nil
	}

	version := record.GetInt("version")
	if version == 0 {
		return echo.NewHTTPError(http.StatusPreconditionRequired, "the version attribute is required for this operation and is missing from the request.")
	}

	err := app.Dao().RunInTransaction(func(txDao *daos.Dao) error {

		collectionName := record.Collection().Name
		id := record.Id
		foundRecord, err := txDao.FindRecordById(collectionName, id)
		if err != nil {
			return echo.NewHTTPError(http.StatusNotFound, "cannot find record with id "+id+" in collection "+collectionName)
		}

		foundVersion := foundRecord.GetInt("version")

		if version != foundVersion {
			return echo.NewHTTPError(http.StatusPreconditionFailed, "the version attribute does not match the current version of the resource. Please reload and try again.")
		}

		record.Set("version", version+1)
		foundRecord.Set("version", version+1)

		if err := txDao.SaveRecord(foundRecord); err != nil {
			return echo.NewHTTPError(http.StatusPreconditionFailed, "the version attribute does not match the current version of the resource. Please reload and try again.")
		}

		return nil
	})

	if err != nil {
		return err
	}

	return nil
})
```
